haskell/test/SMT/Common/Gen.hs

```diff
@@ -0,0 +1,431 @@
+    genTransferContractTokensCall sender1 = do
+      f <- genTransferContractTokens
+      pure $ \govAddr -> mkSimpleContractCall sender1 $ f govAddr
+
+
+mkSimpleContractCall :: Address -> Parameter -> ModelCall
+mkSimpleContractCall sender1 param = do
+  ModelCall
+    { mcAdvanceLevel = Nothing
+    , mcParameter = param
+    , mcSource = ModelSource
+        { msoSource = sender1
+        , msoSender = sender1
+        }
+    }
+
+genProposingProcess :: GeneratorT (ModelInputArg -> [ModelCall])
+genProposingProcess = do
+  userAddrs <- get <&> gsAddresses
+  sender1 <- fst <$> Gen.element userAddrs
+  delegate1 <- fst <$> Gen.element userAddrs
+  invalidFrom <- fst <$> Gen.element userAddrs
+
+  mkGenPropose <- get <&> gsMkGenPropose
+  mkPropose <- mkGenPropose sender1 delegate1 invalidFrom
+
+  mkGenCustomCalls <- get <&> gsMkCustomCalls
+  mkCustomCalls <- mkGenCustomCalls
+
+  -- gen entrypoints
+  flushAction <- genFlush
+  mkVote <- genVote
+  mkDropProposal <- genDropProposal
+
+  -- | `advanceLvl` is provided precisely to ensure the order of operations results in successful proposal.
+  let mkCall param advanceLvl = ModelCall
+        { mcAdvanceLevel = advanceLvl
+        , mcParameter = param
+        , mcSource = ModelSource
+          { msoSource = sender1
+          , msoSender = sender1
+          }
+        }
+
+  pure $ \args@ModelInputArg{..} ->
+    let
+      (proposeAction, metaSize, key) = mkPropose miaGuardianAddr miaGovAddr
+      freezeAmt = metaSize + fixedFee
+      permitProtecteds = mkVote key freezeAmt
+      voterFreezeAmt = getVoteFreezeAmt permitProtecteds
+      dropProposalAction = mkDropProposal key
+    in
+      [ mkCall (XtzForbidden $ Update_delegate [DelegateParam True delegate1]) Nothing
+      , mkCall (XtzForbidden $ Freeze (#amount .! freezeAmt) ) Nothing
+      , mkCall (XtzForbidden $ Freeze (#amount .! voterFreezeAmt) ) Nothing
+      , mkCall proposeAction (Just 1)
+      , mkCall (XtzForbidden $ Vote permitProtecteds) (Just 1)
+      , mkCall flushAction (Just 2)
+      , mkCall (XtzForbidden $ Unfreeze (#amount .! freezeAmt) ) Nothing
+      , mkCall dropProposalAction Nothing
+      ] <> ((\param -> mkCall (XtzAllowed $ CallCustom param) Nothing) <$> mkCustomCalls args)
+  where
+    getVoteFreezeAmt permitProtecteds =
+      foldl' (\i (PermitProtected voteParam _) -> (voteParam & vVoteAmount) + i) 0 permitProtecteds
+
+genVote :: GeneratorT (ProposalKey -> Natural -> [PermitProtected VoteParam])
+genVote = do
+  mkGenVoteParams <- vectorOf 5 $ genVoteParams
+  pure $ \proposalKey correctVoteAmt -> (\f -> f proposalKey correctVoteAmt) <$> mkGenVoteParams
+  where
+    genVoteParams :: GeneratorT (ProposalKey -> Natural -> PermitProtected VoteParam)
+    genVoteParams = do
+      userPool <- get <&> gsAddresses
+      sender_ <- Gen.element userPool
+
+      invalidFrom <- Gen.element userPool
+      from <- Gen.element [sender_, invalidFrom]
```

Comment:
Isn't this the same as `from <- Gen.element userPool`? What makes `invalidFrom` invalid?

Reply:
Yeah, the `invalidFrom` does not make sense in here, I've removed this.